design/kanister-prometheus-integration.md

```diff
@@ -0,0 +1,290 @@
+we would create the bounded labels in the following way:
+
+##### BoundedLabel example
+```golang
+BoundedLabel {
+  LabelName: "operation_type"
+  LabelValues: ["backup", "restore"]
+}
+
+BoundedLabel {
+  LabelName: "action_set_resolution"
+  LabelValues: ["success", "failure"]
+}
+``` 
+##### Initialization methods
+```golang
+ 
+// InitCounterVec initializes and registers the counter metrics vector. It takes a list of 
+// BoundedLabel objects - if any label value or label name is nil, then this method will panic. 
+// Based on the combinations returned by generateCombinations, it will set each counter value to 0.
+// If a nil counter is returned during registration, the method will
+// panic
+func InitCounterVec(r prometheus.Registerer, opts prometheus.CounterOpts, boundedLabels []BoundedLabel) *prometheus.CounterVec
+
+// InitGaugeVec initializes the gauge metrics vector. It takes a list of BoundedLabels, but the 
+// LabelValue field of each BoundedLabel will be ignored.
+// If a nil counter is returned during registration, the method will
+// panic
+func InitGaugeVec(r prometheus.Registerer, opts prometheus.CounterOpts, boundedLabels []BoundedLabel) *prometheus.GaugeVec
+
+// InitHistogramVec initializes the histogram metrics vector. It takes a list of BoundedLabels, but the 
+// LabelValue field of each BoundedLabel will be ignored.
+// If a nil counter is returned during registration, the method will
+// panic
+func InitHistogramVec(r prometheus.Registerer, opts prometheus.CounterOpts, boundedLabels []BoundedLabel) *prometheus.HistogramVec
+
+// InitCounter initializes a new counter.
+// If a nil counter is returned during registration, the method will
+// panic
+func InitCounter(r prometheus.Registerer, opts prometheus.CounterOpts) prometheus.Counter
+
+// InitGauge initializes a new gauge.
+// If a nil counter is returned during registration, the method will
+// panic
+func InitGauge(r prometheus.Registerer, opts prometheus.GaugeOpts) prometheus.Gauge
+
+// InitHistogram initializes a new histogram.
+// If a nil counter is returned during registration, the method will
+// panic
+func InitHistogram(r prometheus.Registerer, opts prometheus.HistogramOpts) prometheus.Histogram
+```
+
+##### Example Initialization Steps for a new CounterVec metric
+
+1. Initialize a new CounterVec with relevant options and label names 
+
+2. Attempt to register the new CounterVec
+
+   a. If successful,
+
+       i. Generate combinations of label names
+
+       ii. Create counters for each combination and set the counter 
+           to 0.
+
+   b. If not successful, check if the error is an AlreadyRegisteredError
+
+       i. If yes, return the CounterVec and ignore the error
+
+       ii. If no, return a nil CounterVec and the received error. 
+
+3. If received a nil CounterVec from registration, interrupt with a panic, 
+   because an interrupt would suggest a failure in the created 
+   CounterVec, which should be fixed by the programmer. 
```

Comment:
```suggestion
1. Initialize a new `CounterVec` with relevant options and label names.

2. Attempt to register the new `CounterVec`.

   a. If successful,

       i. Generate combinations of label names

       ii. Create counters for each combination and set the counter to 0

   b. If not successful, check if the error is an `AlreadyRegisteredError`

       i. If yes, return the `CounterVec` and ignore the error

       ii. If no, return a nil `CounterVec` and the received error

3. If received a nil `CounterVec` from registration, interrupt with a panic,
   because an interrupt would suggest a failure in the created `CounterVec`,
   which should be fixed by the programmer.
```